core-processor/src/ext.rs

```diff
@@ -506,21 +446,47 @@ impl<'a, LP: LazyPagesInterface> ExtMutator<'a, LP> {
         Ok(())
     }
 
-    fn charge_expiring_resources<T: Packet>(
-        &mut self,
-        packet: &T,
-        check_gas_limit: bool,
-    ) -> Result<(), FallibleExtError> {
-        let gas_limit = if check_gas_limit {
-            self.check_gas_limit(packet.gas_limit())?
-        } else {
-            packet.gas_limit().unwrap_or(0)
-        };
+    fn charge_expiring_resources<T: Packet>(&mut self, packet: &T) -> Result<(), FallibleExtError> {
+        let reducing_gas_limit = self.get_reducing_gas_limit(packet)?;
 
-        self.reduce_gas(gas_limit)?;
+        self.reduce_gas(reducing_gas_limit)?;
         self.charge_message_value(packet.value())
     }
 
+    fn get_reducing_gas_limit<T: Packet>(&self, packet: &T) -> Result<u64, FallibleExtError> {
+        match T::kind() {
+            DispatchKind::Handle => {
+                // Any "handle" gasless and gasful *non zero* message must
+                // cover mailbox threshold. That's because destination
+                // of the message is unknown, so it could be a user,
+                // and if gasless message is sent, there must be a
+                // guaranteed gas to cover mailbox.
+                let mailbox_threshold = self.context.mailbox_threshold;
+                let gas_limit = packet.gas_limit().unwrap_or(mailbox_threshold);
+
+                // Zero gasful message is a special case.
+                if gas_limit != 0 && gas_limit < mailbox_threshold {
+                    return Err(MessageError::InsufficientGasLimit.into());
+                }
+
+                Ok(gas_limit)
+            }
+            DispatchKind::Init | DispatchKind::Reply => {
```

Comment:
what if we create program from code_id and salt and code_id doesnt exist, so `store_new_programs` fails and this init message will be treated as user message, isn't it?

needs double checking how we handle it

Reply:
discussed